src/Catalyst.Core.Modules.Dfs/UnixFs/SizeChunker.cs

```diff
@@ -0,0 +1,199 @@
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Catalyst.Abstractions.Dfs.CoreApi;
+using Catalyst.Abstractions.Keystore;
+using Catalyst.Abstractions.Options;
+using Catalyst.Core.Lib.Dag;
+using ProtoBuf;
+
+namespace Catalyst.Core.Modules.Dfs.UnixFs
+{
+    /// <summary>
+    ///     Chunks a data stream into data blocks based upon a size.
+    /// </summary>
+    public class SizeChunker
+    {
+        /// <summary>
+        ///     Performs the chunking.
+        /// </summary>
+        /// <param name="stream">
+        ///     The data source.
+        /// </param>
+        /// <param name="name">
+        ///     A name for the data.
+        /// </param>
+        /// <param name="options">
+        ///     The options when adding data to the IPFS file system.
+        /// </param>
+        /// <param name="blockService">
+        ///     The destination for the chunked data block(s).
+        /// </param>
+        /// <param name="keyChain">
+        ///     Used to protect the chunked data blocks(s).
+        /// </param>
+        /// <param name="cancel">
+        ///     Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException" /> is raised.
+        /// </param>
+        /// <returns>
+        ///     A task that represents the asynchronous operation. The task's value is
+        ///     the sequence of file system nodes of the added data blocks.
+        /// </returns>
+        public async Task<List<UnixFsNode>> ChunkAsync(Stream stream,
+            string name,
+            AddFileOptions options,
+            IBlockApi blockService,
+            IKeyApi keyChain,
+            CancellationToken cancel)
+        {
+            var protecting = !string.IsNullOrWhiteSpace(options.ProtectionKey);
+            var nodes = new List<UnixFsNode>();
+            var chunkSize = options.ChunkSize;
+            var chunk = new byte[chunkSize];
+            var chunking = true;
+            var totalBytes = 0UL;
+
+            while (chunking)
+            {
+                // Get an entire chunk.
+                var length = 0;
+                while (length < chunkSize)
+                {
+                    var n = await stream.ReadAsync(chunk, length, chunkSize - length, cancel).ConfigureAwait(false);
+                    if (n < 1)
+                    {
+                        chunking = false;
+                        break;
+                    }
+
+                    length += n;
+                    totalBytes += (uint) n;
+                }
+
+                //  Only generate empty block, when the stream is empty.
+                if (length == 0 && nodes.Count > 0)
+                {
+                    chunking = false;
+                    break;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor the code in order to remove this break statement.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)